bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/service/RemoteControllerService.java

```diff
@@ -253,264 +169,446 @@ public boolean isUpnp() {
-                    int num4 = val / 1000 % 10;
-                    int num3 = val / 100 % 10;
-                    int num2 = val / 10 % 10;
-                    int num1 = val % 10;
-
-                    List<KeyCode> commands = new ArrayList<>();
-
-                    if (num4 > 0) {
-                        commands.add(KeyCode.valueOf("KEY_" + num4));
-                    }
-                    if (num4 > 0 || num3 > 0) {
-                        commands.add(KeyCode.valueOf("KEY_" + num3));
-                    }
-                    if (num4 > 0 || num3 > 0 || num2 > 0) {
-                        commands.add(KeyCode.valueOf("KEY_" + num2));
-                    }
-                    commands.add(KeyCode.valueOf("KEY_" + num1));
+                    KeyCode[] codes = String.valueOf(decimalCommand.intValue()).chars()
+                            .mapToObj(c -> KeyCode.valueOf("KEY_" + String.valueOf((char) c))).toArray(KeyCode[]::new);
+                    List<Object> commands = new ArrayList<>(Arrays.asList(codes));
                     commands.add(KeyCode.KEY_ENTER);
-                    sendKeyCodes(commands);
-                } else {
-                    logger.warn("Remote control: unsupported command type {} for channel {}", command, channel);
+                    sendKeys(commands);
+                    result = true;
                 }
-                return;
+                break;
             default:
-                logger.warn("Remote control: unsupported channel: {}", channel);
+                logger.warn("{}: Remote control: unsupported channel: {}", host, channel);
+                return false;
+        }
+        if (!result) {
+            logger.warn("{}: Remote control: wrong command type {} channel {}", host, command, channel);
         }
+        return result;
     }
 
-    /**
-     * Sends a command to Samsung TV device.
-     *
-     * @param key Button code to send
-     */
-    private void sendKeyCode(KeyCode key) {
-        try {
-            if (remoteController != null) {
-                remoteController.sendKey(key);
-            }
-        } catch (RemoteControllerException e) {
-            reportError(String.format("Could not send command to device on %s:%d", host, port), e);
-        }
+    public synchronized void sendKeys(KeyCode key, int press) {
+        sendKeys(Arrays.asList(key), press);
     }
 
-    private void sendKeyCodePress(KeyCode key) {
-        try {
-            if (remoteController instanceof RemoteControllerWebSocket remoteControllerWebSocket) {
-                remoteControllerWebSocket.sendKeyPress(key);
-            }
-        } catch (RemoteControllerException e) {
-            reportError(String.format("Could not send command to device on %s:%d", host, port), e);
-        }
+    public synchronized void sendKeys(KeyCode key) {
+        sendKeys(Arrays.asList(key), 0);
+    }
+
+    public synchronized void sendKeys(List<Object> keys) {
```

Comment:
Because they can be sent automatically (like KEY_POWER), or by the user sending a command, or a command sequence. If they aren't synchronized, you can get keys interleaved with other keys, and that destroys the sequencing and timing.